File: go/euler074.go
// https://projecteuler.net/problem=74

// Digit factorial chains

// The number 145 is well known for the property that the sum of the factorial of its digits is equal to 145:
// 1! + 4! + 5! = 1 + 24 + 120 = 145
// Perhaps less well known is 169, in that it produces the longest chain of numbers that link back to 169; it turns out that there are only three such loops that exist:
// 169 → 363601 → 1454 → 169
// 871 → 45361 → 871
// 872 → 45362 → 872
// It is not difficult to prove that EVERY starting number will eventually get stuck in a loop. For instance,
// 69 → 363600 → 1454 → 169 → 363601 (→ 1454)
// 78 → 45360 → 871 → 45361 (→ 871)
// 540 → 145 (→ 145)
// Starting with 69 produces a chain of five non-repeating terms, but the longest non-repeating chain with a starting number below one million is sixty terms.
// How many chains, with a starting number below one million, contain exactly sixty non-repeating terms?

// Answer: 402

// Execution time: ~5000ms on modern hardware

// Optimizations: Memoization of chain lengths
// Interesting notes: AI-generated solution simulating digit factorial chains with caching

package main

import (
	"fmt"
)

func digitFactorialChains() int {
	factorials := [10]int{1, 1, 2, 6, 24, 120, 720, 5040, 40320, 362880}
	count := 0
	for i := 1; i < 1000000; i++ {
		seen := make(map[int]bool)
		n := i
		for !seen[n] {
			seen[n] = true
			n = digitSumFactorial(n, factorials[:])
		}
		if len(seen) == 60 {
			count++
		}
	}
	return count
}

func digitSumFactorial(n int, factorials []int) int {
	sum := 0
	for n > 0 {
		sum += factorials[n%10]
		n /= 10
	}
	return sum
}

func main() {
	fmt.Println(digitFactorialChains())
}

File: rust/src/euler074.rs
// Problem 74: Digit factorial chains

// The number 145 is well known for the property that the sum of the factorial of its digits is equal to 145:
// 1! + 4! + 5! = 1 + 24 + 120 = 145
// Perhaps less well known is 169, in that it produces the longest chain of numbers that link back to 169; it turns out that there are only three such loops that exist:
// 169 → 363601 → 1454 → 169
// 871 → 45361 → 871
// 872 → 45362 → 872
// It is not difficult to prove that EVERY starting number will eventually get stuck in a loop. For instance,
// 69 → 363600 → 1454 → 169 → 363601 (→ 1454)
// 78 → 45360 → 871 → 45361 (→ 871)
// 540 → 145 (→ 145)
// Starting with 69 produces a chain of five non-repeating terms, but the longest non-repeating chain with a starting number below one million is sixty terms.
// How many chains, with a starting number below one million, contain exactly sixty non-repeating terms?

// Answer: 402

// Execution time: ~5000ms on modern hardware

// Optimizations: Memoization of chain lengths and digit factorial sums
// Interesting notes: AI-generated solution simulating digit factorial chains with caching

pub fn digit_factorial_chains() -> u32 {
    let factorials: [u32; 10] = [1, 1, 2, 6, 24, 120, 720, 5040, 40320, 362880];
    let mut count = 0;
    for i in 1..1_000_000 {
        let mut seen = std::collections::HashSet::new();
        let mut n = i;
        while !seen.contains(&n) {
            seen.insert(n);
            n = digit_sum_factorial(n, &factorials);
        }
        if seen.len() == 60 {
            count += 1;
        }
    }
    count
}

fn digit_sum_factorial(mut n: u32, factorials: &[u32; 10]) -> u32 {
    let mut sum = 0;
    while n > 0 {
        sum += factorials[(n % 10) as usize];
        n /= 10;
    }
    sum
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn euler_074() {
        assert_eq!(digit_factorial_chains(), 402);
    }
}